File: Data Analyst in R/Step 3 - Data Cleaning in R/2. Data Cleaning in R Advanced/1. Regular Expression Basics.r
## 1. Introduction ##

library(readr)
hn  <- read_csv("hacker_news.csv")

## 2. The Regular Expression Functions in R ##

library(stringr)
titles <- hn$title
library(dplyr)
pattern <- "Amazon"
matches <- str_detect(titles, pattern)
hn_matches <- if_else(matches, "Match", "No Match")

## 3. Set of characters in Regular Expression ##

pattern <- "[Aa]mazon"
matches <- str_detect(titles, pattern)
amazon_mentions <- sum(matches)

## 4. Alternative patterns ##

pattern <- "2000|2005|2010"
matches <- str_detect(titles, pattern)
hn_matches <- if_else(matches, "Match", "No Match")
hn_group <- hn %>%
    mutate(year_group = hn_matches)

## 5. Using Regular Expressions to Select Data ##

google_titles_logical  <- str_detect(titles, "[Gg]oogle" )
google_titles  <-  titles[google_titles_logical]
hn_google <- hn %>%
    filter(google_titles_logical)

## 6. Quantifiers ##

# The `titles` variable is available from
# previous screens
email_logical  <-  str_detect(titles, "e-?mail")
email_count  <-  sum(email_logical)
email_titles  <-  titles[email_logical]

## 7. Character Classes ##

pattern  <-  "\\[\\w+\\]"
tag_logical  <-  str_detect(titles, pattern)
tag_titles  <-  titles[tag_logical]
tag_count  <-  sum(tag_logical)

## 8. Accessing the Matching Text with Capture Groups ##

tag_titles  <-  titles[str_detect(titles, "\\[\\w+\\]")]
#the previous pattern  is "\\[\\w+\\]"
pattern  <-  "\\[(\\w+)\\]"
tags_text_matches <- str_match(tag_titles, pattern)[,2]
tags_freq <- table(tags_text_matches)

## 9. Negative Character Classes ##

first_10_matches <- function (data, pattern) {
    matches <- str_detect(data, pattern) # finding pattern matches
    matched_df <- data[matches] # subsetting data (keep only matches)
    head(matched_df, 10) # taking the first ten matched elements
}
pattern  <-  "[Jj]ava[^Ss]"
java_titles  <-  titles[str_detect(titles, pattern)]

## 10. Word Boundaries ##

first_10_matches <- function (data, pattern) {
    matches <- str_detect(data, pattern) # finding pattern matches
    matched_df <- data[matches] # subsetting data (keep only matches)
    head(matched_df, 10) # taking the first ten matched elements
}
pattern  <-  "\\b[Jj]ava\\b"
java_titles  <-  titles[str_detect(titles, pattern)]

## 11. Matching at the Start and End of Strings ##

pattern_beginning  <-  "^\\[\\w+\\]"
beginning_count  <-  sum(str_detect(titles, pattern_beginning))

pattern_ending  <-  "\\[\\w+\\]$"
ending_count  <-  sum(str_detect(titles, pattern_ending))

## 12. Challenge: Using Flags to Modify Regex Patterns ##

email_tests  <-  c('email', 'Email', 'e Mail', 'e mail', 'E-mail',
                   'e-mail', 'eMail', 'E-Mail', 'EMAIL')
pattern <- "(?i)e[\\-\\s]?mail"
email_tests_matches <- str_detect(email_tests, pattern)
email_mentions <- sum(str_detect(titles, pattern))

File: Data Scientist In Python/Step 2 - Data Analysis and Visualization/5. Data Cleaning in Python Advanced/1. Regular Expression Basics.py
## 1. Introduction ##

import pandas as pd
hn = pd.read_csv("hacker_news.csv")

## 2. The Regular Expression Module ##

import re

titles = hn["title"].tolist()
python_mentions = 0
pattern = "[Pp]ython"

for t in titles:
    if re.search(pattern, t):
        python_mentions += 1

## 3. Counting Matches with pandas Methods ##

pattern = '[Pp]ython'
titles = hn['title']
python_mentions = titles.str.contains(pattern).sum()

## 4. Using Regular Expressions to Select Data ##

titles = hn['title']
ruby_titles = titles[titles.str.contains(r"[Rr]uby")]

## 5. Quantifiers ##

# The `titles` variable is available from
# the previous screens
email_bool = titles.str.contains("e-?mail")
email_count = email_bool.sum()
email_titles = titles[email_bool]

## 6. Character Classes ##

pattern = "\[\w+\]"
tag_titles = titles[titles.str.contains(pattern)]
tag_count = tag_titles.shape[0]

## 7. Accessing the Matching Text with Capture Groups ##

# pattern = r"\[\w+\]"
pattern = r"\[(\w+)\]"
tag_freq = titles.str.extract(pattern).value_counts()

## 8. Negative Character Classes ##

def first_10_matches(pattern):
    """
    Return the first 10 story titles that match
    the provided regular expression
    """
    all_matches = titles[titles.str.contains(pattern)]
    first_10 = all_matches.head(10)
    return first_10
pattern = r"[Jj]ava[^Ss]"
java_titles = titles[titles.str.contains(pattern)]

## 9. Word Boundaries ##

pattern = r"\b[Jj]ava\b"
java_titles = titles[titles.str.contains(pattern)]

## 10. Matching at the Start and End of Strings ##

pattern_beginning = r"^\[\w+\]"
beginning_count = titles.str.contains(pattern_beginning).sum()

pattern_ending =  r"\[\w+\]$"
ending_count = titles.str.contains(pattern_ending).sum()

## 11. Challenge: Using Flags to Modify Regex Patterns ##

import re

email_tests = pd.Series(['email', 'Email', 'e Mail', 'e mail', 'E-mail',
              'e-mail', 'eMail', 'E-Mail', 'EMAIL', 'emails', 'Emails',
              'E-Mails'])
pattern = r"\be[\-\s]?mails?\b"
email_mentions = titles.str.contains(pattern, flags=re.I).sum()